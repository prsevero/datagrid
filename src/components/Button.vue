<template>
    <button :type="type">{{ label }}</button>
</template>


<script>
export default {
    name: 'Button',
    props: {
        label: {
            required: true,
            type: String,
        },
        type: {
            default: 'button',
            type: String,
        },
    },
};
</script>


<style lang="sass" scoped>
button
    background: #07b
    border: 0
    border-radius: 0
    color: #fff
    cursor: pointer
    display: block
    margin: 0 auto
    outline: 0
    padding: 10px 12px
    transition: all .22s ease-in-out
    &:focus, &:hover
        background: #08c
</style>
