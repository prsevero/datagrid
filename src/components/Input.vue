<template>
    <label>
        <input :class="{'input-error': hasError}" :required="required" :type="type" :value="value" @input="$emit('input', $event.target.value)">
        <span>{{ label }}</span>
    </label>
</template>


<script>
export default {
    name: 'Input',
    props: {
        hasError: Boolean,
        label: {
            required: true,
            type: String,
        },
        required: Boolean,
        type: {
            required: true,
            type: String,
        },
        value: [String, Number],
    },
};
</script>

<style lang="sass" scoped>
label
    display: block
    margin: 0 0 20px
    position: relative
    span
        background: #fff
        color: #333
        cursor: text
        left: 5px
        padding: 0 4px
        pointer-events: none
        position: absolute
        top: 12px
        transition: all .15s ease-in-out

input
    appearance: none
    border: 1px solid #ccc
    border-radius: 0
    color: #333
    display: block
    outline: 0
    padding: 8px
    width: 100%
    transition: all .22s ease-in-out
    &.input-error
        border-color: #f00
    &:focus
        border-color: #777
    &:focus, &:valid
        ~ span
            color: #444
            font-size: .8em
            top: -9px
</style>
