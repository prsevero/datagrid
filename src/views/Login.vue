<template>
    <div id="login">
        <form novalidate @submit.prevent="onSubmit">
            <img alt="Company Name Logo" src="../assets/logo.svg" />
            <Input label="Username" type="text" :has-error="!hasUsername" :required="true" v-model="username" />
            <Input label="Password" type="password" :has-error="!hasPassword" :required="true" v-model="password" />
            <Button label="Enter" type="submit" />
        </form>
    </div>
</template>


<script>
import Button from '@/components/Button.vue';
import Input from '@/components/Input.vue';

export default {
    name: 'login',
    components: {Button, Input},
    data() {
        return {
            password: '',
            submitted: false,
            username: '',
        };
    },
    computed: {
        hasPassword() {
            return !this.submitted || this.password !== '';
        },
        hasUsername() {
            return !this.submitted || this.username !== '';
        },
    },
    methods: {
        isLoggedIn() {
            return localStorage.getItem('token');
        },
        login(token) {
            localStorage.setItem('token', token);
        },
        logout() {
            localStorage.removeItem('token');
        },
        onSubmit() {
            this.submitted = true;
            if (this.username && this.password) {
                this.login('SAMPLE_TOKEN');
                this.$router.replace({path: '/'});
            }
        },
    },
    mounted() {
        if (this.$router.history.current.redirectedFrom === '/logout')
            this.logout();
        else if (this.isLoggedIn()) this.$router.replace({path: '/'});
    },
};
</script>


<style lang="sass" scoped>
#login
    align-items: center
    bottom: 100px
    display: flex
    left: 0
    justify-content: center
    position: absolute
    right: 0
    top: 0

form
    flex: 1 0 0
    font-size: 1.6em
    max-width: 320px
    padding: 0 15px

    img
        display: block
        margin: 0 auto 20px
        max-width: 75px
</style>
