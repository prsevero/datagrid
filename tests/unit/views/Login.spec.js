import { createLocalVue, shallowMount } from '@vue/test-utils';
import VueRouter from 'vue-router';
import Button from '@/components/Button.vue';
import Input from '@/components/Input.vue';
import Login from '@/views/Login.vue';


const localVue = createLocalVue();
localVue.use(VueRouter);
const router = new VueRouter();


describe('Login.vue', () => {

    let mountedLogin;
    const login = () => {
        if (!mountedLogin) {
            mountedLogin = shallowMount(Login, {localVue, router});
        }
        return mountedLogin;
    }

    beforeEach(() => {
        localStorage.removeItem('token');
        mountedLogin = undefined;
    });

    it('should match snapshot', () => {
        const wrapper = login();
        expect(wrapper.vm.$el).toMatchSnapshot();
    });

    it('always render login div', () => {
        const wrapper = login();
        expect(wrapper.element.id).toMatch('login');
        expect(wrapper.is('div')).toBe(true);
    });

    it('always render a form', () => {
        const wrapper = login();
        expect(wrapper.findAll('form').length).toBe(1);
    });

    it('always render an image', () => {
        const wrapper = login();
        expect(wrapper.findAll('img').length).toBe(1);
    });

    it('always render the username and the password inputs', () => {
        const wrapper = login();
        const inputs = wrapper.findAll(Input);
        expect(inputs.length).toBe(2);
        expect(inputs.at(0).attributes().label).toBe('Username');
        expect(inputs.at(1).attributes().label).toBe('Password');
    });

    it('always render the submit button', () => {
        const wrapper = login();
        const button = wrapper.find(Button);
        expect(button.exists());
        expect(button.attributes().type).toBe('submit');
    });

    it('does not set localStorage token if the credentials are invalid', () => {
        const wrapper = login();
        const button = wrapper.find(Button);
        button.trigger('submit');
        expect(localStorage.getItem('token')).toBeNull();
    });

    describe('when the credentials are valid', () => {
        it('sets localStorage token', () => {
            const wrapper = login();
            const button = wrapper.find(Button);
            wrapper.setData({username: 'username', password: 'password'});
            button.trigger('submit');
            expect(localStorage.getItem('token')).toBe('SAMPLE_TOKEN');
        });

        it('redirects to home', () => {
            const wrapper = login();
            const button = wrapper.find(Button);
            wrapper.setData({username: 'username', password: 'password'});
            button.trigger('submit');
            expect(router.history.current.path).toBe('/');
        });
    });

    it('removes the token from localStorage when logouts', () => {
        const wrapper = login();
        wrapper.vm.login('SAMPLE_TOKEN');
        expect(localStorage.getItem('token')).toBe('SAMPLE_TOKEN');
        wrapper.vm.logout();
        expect(localStorage.getItem('token')).toBeNull();
    });

});
