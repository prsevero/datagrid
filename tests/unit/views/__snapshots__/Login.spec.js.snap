// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue should match snapshot 1`] = `
<div
  id="login"
>
  <form
    novalidate="novalidate"
  >
    <img
      alt="Company Name Logo"
      src="../assets/logo.svg"
    />
     
    <input-stub
      label="Username"
      required="true"
      type="text"
      value=""
    />
     
    <input-stub
      label="Password"
      required="true"
      type="password"
      value=""
    />
     
    <button-stub
      label="Enter"
      type="submit"
    />
  </form>
</div>
`;
